File: pipeline/derive.py
#!/usr/bin/env python3
"""
Deterministic derivation of status, health score, and tags from raw scan data.

Reads raw scan JSON from stdin, outputs enriched JSON to stdout.

Status rules (by daysInactive):
  - active:   <= 14 days
  - paused:   15-60 days
  - stale:    61-180 days
  - archived: > 180 days or no commits

Hygiene score (structural, slow-moving, 0-95 raw -> 0-100):
  - README: +15, Tests: +20, CI/CD: +15, Remote: +10
  - Low TODOs (<10): +10, Deployment: +10, Linter: +5, License: +5, Lockfile: +5

Momentum score (operational, fast-moving, 0-70 raw -> 0-100):
  - Commit recency: +25 (<=7d), +20 (<=14d), +15 (<=30d), +5 (<=60d)
  - Clean working tree (!isDirty): +20
  - Pushed up (ahead==0): +15
  - Low stale branches (<=3): +10

Health score (legacy, backward compatible):
  - round(0.65 * hygiene + 0.35 * momentum)
"""

import json
import sys


def derive_status(days_inactive: int | None) -> str:
    if days_inactive is None:
        return "archived"
    if days_inactive <= 14:
        return "active"
    if days_inactive <= 60:
        return "completed"
    if days_inactive <= 180:
        return "paused"
    return "archived"


def derive_hygiene_score(project: dict) -> tuple[int, dict[str, int]]:
    """Structural health signals (0-95 raw, normalized to 0-100)."""
    breakdown: dict[str, int] = {}
    files = project.get("files", {})
    cicd = project.get("cicd", {})
    deployment = project.get("deployment", {})

    if files.get("readme"):
        breakdown["readme"] = 15
    if files.get("tests"):
        breakdown["tests"] = 20
    if any(cicd.values()):
        breakdown["cicd"] = 15
    if project.get("remoteUrl"):
        breakdown["remote"] = 10
    if project.get("todoCount", 0) < 10:
        breakdown["lowTodos"] = 10
    if any(deployment.values()):
        breakdown["deployment"] = 10
    if files.get("linterConfig"):
        breakdown["linter"] = 5
    if files.get("license"):
        breakdown["license"] = 5
    if files.get("lockfile"):
        breakdown["lockfile"] = 5

    raw = sum(breakdown.values())
    normalized = min(round(raw * 100 / 95), 100)
    return normalized, breakdown


def derive_momentum_score(project: dict) -> tuple[int, dict[str, int]]:
    """Operational velocity signals (0-70 raw, normalized to 0-100)."""
    breakdown: dict[str, int] = {}

    # Commit recency
    days = project.get("daysInactive")
    if days is not None:
        if days <= 7:
            breakdown["recency"] = 25
        elif days <= 14:
            breakdown["recency"] = 20
        elif days <= 30:
            breakdown["recency"] = 15
        elif days <= 60:
            breakdown["recency"] = 5

    # Clean working tree
    if not project.get("isDirty", False):
        breakdown["cleanTree"] = 20

    # Pushed up (no commits ahead of remote)
    if project.get("ahead", 0) == 0:
        breakdown["pushedUp"] = 15

    # Low stale branches
    if project.get("branchCount", 0) <= 3:
        breakdown["lowBranches"] = 10

    raw = sum(breakdown.values())
    normalized = min(round(raw * 100 / 70), 100)
    return normalized, breakdown


def derive_health_score(project: dict) -> tuple[int, int, int, dict]:
    """Combined scoring: hygiene + momentum -> legacy healthScore."""
    hygiene, hygiene_breakdown = derive_hygiene_score(project)
    momentum, momentum_breakdown = derive_momentum_score(project)
    health = round(0.65 * hygiene + 0.35 * momentum)

    score_breakdown = {
        "hygiene": hygiene_breakdown,
        "momentum": momentum_breakdown,
    }

    return health, hygiene, momentum, score_breakdown


def derive_tags(project: dict) -> list[str]:
    tags: list[str] = []
    languages = project.get("languages", {})
    files = project.get("files", {})
    cicd = project.get("cicd", {})
    deployment = project.get("deployment", {})

    # Language-based tags
    for lang in languages.get("detected", []):
        normalized = lang.lower().replace("/", "-")
        tags.append(normalized)

    # Capability-based tags
    if files.get("dockerfile") or files.get("dockerCompose"):
        tags.append("docker")
    if any(cicd.values()):
        tags.append("ci-cd")
    if any(deployment.values()):
        tags.append("deployed")
    if files.get("tests"):
        tags.append("tested")

    # Framework-based tags (from scan's framework field)
    framework = project.get("framework")
    if framework:
        tags.append(framework.lower())

    # Service-based tags (from scan's services field)
    for service in project.get("services", []):
        tags.append(service.lower())

    return sorted(set(tags))


def derive_project(project: dict) -> dict:
    days_inactive = project.get("daysInactive")
    health, hygiene, momentum, score_breakdown = derive_health_score(project)
    return {
        "pathHash": project["pathHash"],
        "statusAuto": derive_status(days_inactive),
        "healthScoreAuto": health,
        "hygieneScoreAuto": hygiene,
        "momentumScoreAuto": momentum,
        "scoreBreakdownJson": score_breakdown,
        "tags": derive_tags(project),
    }


def main() -> None:
    raw = json.load(sys.stdin)
    projects = raw.get("projects", [])

    derived = [derive_project(p) for p in projects]

    output = {
        "derivedAt": raw.get("scannedAt"),
        "projects": derived,
    }

    print(json.dumps(output, indent=2))


if __name__ == "__main__":
    main()

File: src/lib/pipeline-native/derive.ts
/**
 * TypeScript-native derive — deterministic derivation of status, health score, and tags.
 *
 * Port of pipeline/derive.py. Must produce byte-identical output for the same input.
 * All thresholds and weights are loaded from derive-config.ts for tunability.
 */

import { deriveConfig, hygieneRawMax, momentumRawMax } from "@/config/derive-config";

export interface ScanProject {
  pathHash: string;
  daysInactive: number | null;
  isDirty: boolean;
  ahead: number;
  branchCount: number;
  remoteUrl: string | null;
  todoCount: number;
  framework: string | null;
  languages: { primary: string | null; detected: string[] };
  files: {
    readme: boolean;
    tests: boolean;
    dockerfile: boolean;
    dockerCompose: boolean;
    linterConfig: boolean;
    license: boolean;
    lockfile: boolean;
  };
  cicd: Record<string, boolean>;
  deployment: Record<string, boolean>;
  services: string[];
  [key: string]: unknown;
}

export interface DeriveProject {
  pathHash: string;
  statusAuto: "active" | "completed" | "paused" | "archived";
  healthScoreAuto: number;
  hygieneScoreAuto: number;
  momentumScoreAuto: number;
  scoreBreakdownJson: {
    hygiene: Record<string, number>;
    momentum: Record<string, number>;
  };
  tags: string[];
}

export interface DeriveOutput {
  derivedAt: string;
  projects: DeriveProject[];
}

export function deriveStatus(daysInactive: number | null): DeriveProject["statusAuto"] {
  if (daysInactive === null) return "archived";
  const { active, completed, paused } = deriveConfig.status;
  if (daysInactive <= active) return "active";
  if (daysInactive <= completed) return "completed";
  if (daysInactive <= paused) return "paused";
  return "archived";
}

export function deriveHygieneScore(project: ScanProject): [number, Record<string, number>] {
  const breakdown: Record<string, number> = {};
  const { hygiene } = deriveConfig;
  const files = project.files ?? {};
  const cicd = project.cicd ?? {};
  const deployment = project.deployment ?? {};

  if (files.readme) breakdown.readme = hygiene.readme;
  if (files.tests) breakdown.tests = hygiene.tests;
  if (Object.values(cicd).some(Boolean)) breakdown.cicd = hygiene.cicd;
  if (project.remoteUrl) breakdown.remote = hygiene.remote;
  if ((project.todoCount ?? 0) < deriveConfig.lowTodosThreshold) breakdown.lowTodos = hygiene.lowTodos;
  if (Object.values(deployment).some(Boolean)) breakdown.deployment = hygiene.deployment;
  if (files.linterConfig) breakdown.linter = hygiene.linter;
  if (files.license) breakdown.license = hygiene.license;
  if (files.lockfile) breakdown.lockfile = hygiene.lockfile;

  const raw = Object.values(breakdown).reduce((a, b) => a + b, 0);
  const normalized = Math.min(Math.round((raw * 100) / hygieneRawMax), 100);
  return [normalized, breakdown];
}

export function deriveMomentumScore(project: ScanProject): [number, Record<string, number>] {
  const breakdown: Record<string, number> = {};
  const { momentum } = deriveConfig;

  const days = project.daysInactive;
  if (days !== null) {
    if (days <= 7) breakdown.recency = momentum.recency7d;
    else if (days <= 14) breakdown.recency = momentum.recency14d;
    else if (days <= 30) breakdown.recency = momentum.recency30d;
    else if (days <= 60) breakdown.recency = momentum.recency60d;
  }

  if (!project.isDirty) breakdown.cleanTree = momentum.cleanTree;
  if ((project.ahead ?? 0) === 0) breakdown.pushedUp = momentum.pushedUp;
  if ((project.branchCount ?? 0) <= deriveConfig.lowBranchesThreshold) breakdown.lowBranches = momentum.lowBranches;

  const raw = Object.values(breakdown).reduce((a, b) => a + b, 0);
  const normalized = Math.min(Math.round((raw * 100) / momentumRawMax), 100);
  return [normalized, breakdown];
}

export function deriveTags(project: ScanProject): string[] {
  const tags: Set<string> = new Set();
  const languages = project.languages ?? { detected: [] };
  const files = project.files ?? {};
  const cicd = project.cicd ?? {};
  const deployment = project.deployment ?? {};

  for (const lang of languages.detected ?? []) {
    tags.add(lang.toLowerCase().replaceAll("/", "-"));
  }

  if (files.dockerfile || files.dockerCompose) tags.add("docker");
  if (Object.values(cicd).some(Boolean)) tags.add("ci-cd");
  if (Object.values(deployment).some(Boolean)) tags.add("deployed");
  if (files.tests) tags.add("tested");

  if (project.framework) tags.add(project.framework.toLowerCase());

  for (const service of project.services ?? []) {
    tags.add(service.toLowerCase());
  }

  return [...tags].sort();
}

export function deriveProject(project: ScanProject): DeriveProject {
  const [hygiene, hygieneBreakdown] = deriveHygieneScore(project);
  const [momentum, momentumBreakdown] = deriveMomentumScore(project);
  const { hygiene: hW, momentum: mW } = deriveConfig.healthWeights;
  const health = Math.round(hW * hygiene + mW * momentum);

  return {
    pathHash: project.pathHash,
    statusAuto: deriveStatus(project.daysInactive),
    healthScoreAuto: health,
    hygieneScoreAuto: hygiene,
    momentumScoreAuto: momentum,
    scoreBreakdownJson: {
      hygiene: hygieneBreakdown,
      momentum: momentumBreakdown,
    },
    tags: deriveTags(project),
  };
}

export function deriveAll(scanOutput: { scannedAt: string; projects: ScanProject[] }): DeriveOutput {
  return {
    derivedAt: scanOutput.scannedAt,
    projects: scanOutput.projects.map(deriveProject),
  };
}